### PR TITLE
client: option to surface connection errors to callers

### PR DESCRIPTION
**Pull request by sethp-nr**
This commit allows blocking clients to receive a more informative error
message than "context deadline exceeded", which is especially helpful in
tracking down persistent client misconfiguration (such as an invalid TLS
certificate, an invalid server that's refusing connections, etc.)

### CODE REVIEW

File: dialoptions.go [hunk 1]
```diff
@@ -298,6 +299,16 @@ func WithBlock() DialOption {
 	})
 }
 
+// WithReturnLastError returns a DialOption which makes the client connection
+// return the last connection error that occurred instead of a context.DeadlineExceeded error.
+// Implies WithBlock()
+func WithReturnLastError() DialOption {
```
Comment (dfawley) [inline dialoptions.go]: Please add an experimental tag for now (see other experimental dial options).

File: dialoptions.go [hunk 2]
```diff
@@ -298,6 +299,16 @@ func WithBlock() DialOption {
 	})
 }
 
+// WithReturnLastError returns a DialOption which makes the client connection
+// return the last connection error that occurred instead of a context.DeadlineExceeded error.
```
Comment (dfawley) [inline dialoptions.go]: Please clarify that this returns the connection error string in addition to the context error, not _instead of_ the context error.

File: dialoptions.go [hunk 3]
```diff
@@ -298,6 +299,16 @@ func WithBlock() DialOption {
 	})
 }
 
+// WithReturnLastError returns a DialOption which makes the client connection
```
Comment (dfawley) [inline dialoptions.go]: Regarding the name, `WithReturnConnectionError` is preferable to us.

**Comment by dfawley** [comment]
Thank you for the PR!

File: clientconn.go [hunk 4]
```diff
@@ -216,7 +216,14 @@ func DialContext(ctx context.Context, target string, opts ...DialOption) (conn *
...
 		select {
 		case <-ctx.Done():
-			conn, err = nil, ctx.Err()
+			switch {
+			case ctx.Err() == err:
+				conn = nil
+			case err == nil || !cc.dopts.returnLastError:
+				conn, err = nil, ctx.Err()
+			default:
+				conn, err = nil, fmt.Errorf("%v: %v", ctx.Err(), err)
```
Comment (timbunce) [inline clientconn.go:225]: Wouldn't it be better for the second `%v` to be `%w` here, so the error is wrapped and thus available for introspection?

Comment (dfawley) [reply clientconn.go:225]: #4155 / #4163

**Pull request merged**